rust/datafusion/src/logical_plan/dfschema.rs

```diff
@@ -0,0 +1,415 @@
+    pub fn empty() -> Self {
+        Self { fields: vec![] }
+    }
+
+    /// Create a new `DFSchema`
+    pub fn new(fields: Vec<DFField>) -> Result<Self> {
+        let mut qualified_names: HashSet<(&str, &str)> = HashSet::new();
+        let mut unqualified_names: HashSet<&str> = HashSet::new();
+        for field in &fields {
+            if let Some(qualifier) = field.qualifier() {
+                if !qualified_names.insert((qualifier, field.name())) {
+                    return Err(DataFusionError::Plan(format!(
+                        "Joined schema would contain duplicate qualified field name '{}'",
+                        field.qualified_name()
+                    )));
+                }
+            } else {
+                if !unqualified_names.insert(field.name()) {
+                    return Err(DataFusionError::Plan(
+                        format!("Joined schema would contain duplicate unqualified field name '{}'",
+                                field.name())
+                    ));
+                }
+            }
+        }
+
+        // check for mix of qualified and unqualified field with same unqualified name
+        // note that we need to sort the contents of the HashSet first so that errors are
+        // deterministic
+        let mut qualified_names: Vec<&(&str, &str)> = qualified_names.iter().collect();
+        qualified_names.sort_by(|a, b| {
+            let a = format!("{}.{}", a.0, a.1);
+            let b = format!("{}.{}", b.0, b.1);
+            a.cmp(&b)
+        });
+        for (qualifier, name) in &qualified_names {
+            if unqualified_names.contains(name) {
+                return Err(DataFusionError::Plan(format!(
+                    "Joined schema would contain qualified field name '{}.{}' \
+                    and unqualified field name '{}' which would be ambiguous",
+                    qualifier, name, name
+                )));
+            }
+        }
+        Ok(Self { fields })
+    }
+
+    /// Create a `DFSchema` from an Arrow schema
+    pub fn from(schema: &Schema) -> Result<Self> {
+        Self::new(
+            schema
+                .fields()
+                .iter()
+                .map(|f| DFField {
+                    field: f.clone(),
+                    qualifier: None,
+                })
+                .collect(),
+        )
+    }
+
+    /// Create a `DFSchema` from an Arrow schema
+    pub fn from_qualified(qualifier: &str, schema: &Schema) -> Result<Self> {
+        Self::new(
+            schema
+                .fields()
+                .iter()
+                .map(|f| DFField {
+                    field: f.clone(),
+                    qualifier: Some(qualifier.to_owned()),
+                })
+                .collect(),
+        )
+    }
+
+    /// Combine two schemas
+    pub fn join(&self, schema: &DFSchema) -> Result<Self> {
+        let mut fields = self.fields.clone();
+        fields.extend_from_slice(schema.fields().as_slice());
```

Comment:
it seems like this will allow duplicated field names if `self` and `schema` have such duplication -- I think we may want to call `Self::new(...)` so we get the error checking

Reply:
Ah, good point, I missed that one.

Reply:
Actually, the very next line here is calling `Self::new` so we are doing the required validation and the tests are covering this.